File: src/pcl-bindings/VectorByte.cpp
#include "export.h"
#include <vector>

using point_t = unsigned char;
using vector_t = std::vector<point_t>;

EXPORT(vector_t*) std_vector_byte_ctor()
{
    return new vector_t();
}

EXPORT(vector_t*) std_vector_byte_ctor_count(size_t count)
{
    return new vector_t(count);
}

EXPORT(void) std_vector_byte_delete(vector_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) std_vector_byte_at(vector_t* ptr, size_t idx, point_t* value)
{
    *value = ptr->at(idx);
}

EXPORT(size_t) std_vector_byte_size(vector_t* ptr)
{
    return ptr->size();
}

EXPORT(void) std_vector_byte_clear(vector_t* ptr)
{
    ptr->clear();
}

EXPORT(void) std_vector_byte_resize(vector_t* ptr, size_t size)
{
    ptr->resize(size);
}

EXPORT(void) std_vector_byte_add(vector_t* ptr, point_t value)
{
    ptr->push_back(value);
}

EXPORT(void) std_vector_byte_insert(vector_t* ptr, ptrdiff_t index, point_t value)
{
    ptr->insert(ptr->begin() + index, value);
}

EXPORT(point_t*) std_vector_byte_data(vector_t* ptr)
{
    return ptr->data();
}

File: src/Pcl.NET/VectorByte.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Pcl.NET
{
    public class VectorByte : Vector<byte>
    {
        public override IntPtr Data
        {
            get
            {
                return Invoke.std_vector_byte_data(_ptr);
            }
        }

        public override long Count
        {
            get
            {
                ThrowIfDisposed();
                return (long)Invoke.std_vector_byte_size(_ptr);
            }
        }

        public unsafe VectorByte(byte[] arr)
        {
            _ptr = Invoke.std_vector_byte_ctor_count((ulong)arr.Length);
            CopyFromArray(arr, DataU);
        }

        internal VectorByte(IntPtr ptr)
        {
            _suppressDispose = true;
            _ptr = ptr;
        }

        public VectorByte()
        {
            _ptr = Invoke.std_vector_byte_ctor();
        }

        public VectorByte(long count)
        {
            _ptr = Invoke.std_vector_byte_ctor_count((ulong)count);
        }

        public override void Add(byte item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_byte_add(_ptr, item);
        }

        public override void At(long index, ref byte value)
        {
            ThrowIfDisposed();
            Invoke.std_vector_byte_at(_ptr, (ulong)index, ref value);
        }

        public override void Clear()
        {
            ThrowIfDisposed();
            Invoke.std_vector_byte_clear(_ptr);
        }

        public override void Insert(long index, byte item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_byte_insert(_ptr, (ulong)index, item);
        }

        public override void Resize(long size)
        {
            ThrowIfDisposed();
            Invoke.std_vector_byte_resize(_ptr, (ulong)size);
        }

        public override IEnumerator<byte> GetEnumerator()
        {
            long count = Count;
            for (var i = 0; i < count; i++)
            {
                yield return this[i];
            }
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.std_vector_byte_delete(ref _ptr);
            }
        }
    }
}
